File: lib/src/utils/apiClient.dart
import 'dart:async';
import 'package:dio/dio.dart';
import 'package:cookie_jar/cookie_jar.dart';
import 'package:dio_cookie_manager/dio_cookie_manager.dart';
import 'package:path_provider/path_provider.dart';
import 'apiConfig.dart';
// Event bus equivalent for window.dispatchEvent
class TrialExpiredEvent {
  final Map<String, dynamic> data;
  TrialExpiredEvent(this.data);
}
final StreamController<TrialExpiredEvent> trialEventController = StreamController<TrialExpiredEvent>.broadcast();

final Dio dioClient = Dio(
  BaseOptions(
    baseUrl: ApiConfig.baseUrl,
    // Equivalent to credentials: "include" for cookies
    extra: {'withCredentials': true},
    headers: {
      'X-Requested-With': 'XMLHttpRequest',
    },
  ),
)..interceptors.add(
  InterceptorsWrapper(
    onResponse: (response, handler) {
      final data = response.data;
      if (data != null && data is Map<String, dynamic>) {
        if (data['subscriptionStatus'] == 'EXPIRED' || data['code'] == 'TRIAL_EXPIRED') {
          trialEventController.add(TrialExpiredEvent(data));
          return handler.reject(
            DioException(
              requestOptions: response.requestOptions,
              response: response,
              error: 'TRIAL_EXPIRED',
            ),
          );
        }
      }
      return handler.next(response);
    },
    onError: (DioException e, handler) {
      if (e.response?.statusCode == 402 && e.response?.data?['code'] == 'TRIAL_EXPIRED') {
         trialEventController.add(TrialExpiredEvent(e.response?.data ?? {}));
      }
      return handler.next(e);
    },
  ),
);

late PersistCookieJar cookieJar;

Future<void> initCookies() async {
  final appDocDir = await getApplicationDocumentsDirectory();
  cookieJar = PersistCookieJar(storage: FileStorage("${appDocDir.path}/.cookies/"));
  dioClient.interceptors.add(CookieManager(cookieJar));
}

// Helper function to mimic `apiFetch`
Future<dynamic> apiFetch(String url, {String method = 'GET', dynamic data}) async {
  try {
    final response = await dioClient.request(
        url,
        data: data,
        options: Options(method: method)
    );
    return response.data;
  } on DioException catch (e) {
    if (e.response?.data != null) {
      final errorData = e.response?.data;
      if (errorData is Map) {
         throw Exception(errorData['message'] ?? errorData['error'] ?? 'Request failed');
      }
      throw Exception('Request failed with status ${e.response?.statusCode}: $errorData');
    }
    throw Exception('API ERROR: ${e.message}');
  }
}

File: lib/src/utils/apiClient.js
// // // export async function apiFetch(url, options = {}) {
// // //   const API_BASE = "http://localhost:5000";
// // //   const token = localStorage.getItem("token"); // 🔥 IMPORTANT

// // //   const res = await fetch(`${API_BASE}${url}`, {
// // //     ...options,
// // //     headers: {
// // //       "Content-Type": "application/json",
// // //       Authorization: token ? `Bearer ${token}` : "",
// // //       ...(options.headers || {}),
// // //     },
// // //   });

// // //   const data = await res.json();

// // //   if (res.status === 402 && data.code === "TRIAL_EXPIRED") {
// // //     window.dispatchEvent(
// // //       new CustomEvent("trial-expired", {
// // //         detail: data.contact,
// // //       })
// // //     );
// // //     return null;
// // //   }

// // //   if (!res.ok) {
// // //     throw new Error(data.message || "Request failed");
// // //   }

// // //   return data;
// // // }




// // export async function apiFetch(url, options = {}) {
// //   const API_BASE = "http://localhost:5000";

// //   const res = await fetch(`${API_BASE}${url}`, {
// //     ...options,

// //     credentials: "include",   // 🔥🔥🔥 THIS IS THE NEW TOKEN

// //     headers: {
// //       "Content-Type": "application/json",
// //       ...(options.headers || {}),
// //     },
// //   });

// //   const data = await res.json();

// //   if (res.status === 402 && data.code === "TRIAL_EXPIRED") {
// //     window.dispatchEvent(
// //       new CustomEvent("trial-expired", {
// //         detail: data.contact,
// //       })
// //     );
// //     return null;
// //   }

// //   if (!res.ok) {
// //     throw new Error(data.message || "Request failed");
// //   }

// //   return data;
// // }


// const API_BASE = "http://localhost:5000";

// export async function apiFetch(url, options = {}) {
//   try {
//     const res = await fetch(`${API_BASE}${url}`, {
//       credentials: "include", // ✅ CRITICAL FOR COOKIES
//       headers: {
//         "Content-Type": "application/json",
//         ...(options.headers || {}),
//       },
//       ...options,
//     });

//     const data = await res.json();

//     if (!res.ok) {
//       throw new Error(data.message || "Request failed");
//     }

//     return data;

//   } catch (err) {
//     console.error("API ERROR:", err.message);
//     throw err;
//   }
// }


// const BASE_URL = "http://localhost:5000";
// const BASE_URL = "https://restaurant-model-backend.onrender.com";
// import apiConfig from "../utils/apiConfig";


// export async function apiFetch(url, options = {}) {
//   try {
//     const res = await fetch(`${apiConfig?.BASE_URL}${url}`, {
//       credentials: "include",   // 🔥 REQUIRED FOR COOKIES
//       ...options,
//       headers: {
//         ...(options.body instanceof FormData
//           ? {}  // ❌ Don't set JSON header for FormData
//           : { "Content-Type": "application/json" }),
//         ...(options.headers || {}),
//       },
//     });

//     const data = await res.json().catch(() => ({}));

//     if (!res.ok) {
//       throw new Error(data.message || "Request failed");
//     }

//     return data;

//   } catch (err) {
//     console.error("API ERROR:", err.message);
//     throw err;
//   }
// }



// import apiConfig from "../utils/apiConfig";

// export async function apiFetch(url, options = {}) {
//   try {

//     const res = await fetch(`${apiConfig?.BASE_URL}${url}`, {
//       credentials: "include",
//       ...options,
//       headers: {
//         ...(options.body instanceof FormData
//           ? {}
//           : { "Content-Type": "application/json" }),
//         ...(options.headers || {}),
//       },
//     });

//     const data = await res.json().catch(() => ({}));

//     /*
//      🔥 GLOBAL TRIAL EXPIRED HANDLER
//     */
//     if (data?.code === "TRIAL_EXPIRED") {

//       // Prevent multiple popups
//       if (!window.__trialPopupShown) {
//         window.__trialPopupShown = true;

//         window.dispatchEvent(
//           new CustomEvent("trialExpired")
//         );
//       }

//       return Promise.reject(data);
//     }

//     if (!res.ok) {
//       throw new Error(data.message || "Request failed");
//     }

//     return data;

//   } catch (err) {
//     console.error("API ERROR:", err.message);
//     throw err;
//   }
// }




import apiConfig from "../utils/apiConfig";

export async function apiFetch(url, options = {}) {
  try {

    const res = await fetch(`${apiConfig?.BASE_URL}${url}`, {
      credentials: "include",
      ...options,
      headers: {
        ...(options.body instanceof FormData
          ? {}
          : { "Content-Type": "application/json" }),
        ...(options.headers || {}),
      },
    });

    const data = await res.json().catch(() => ({}));
    console.log(data, "Data");
    /*
     🔥 TRIAL EXPIRED DETECTION
    */
    if (data?.subscriptionStatus === "EXPIRED") {
console.log("inside the condition ")
      if (typeof window !== "undefined") {
        window.dispatchEvent(new Event("TRIAL_EXPIRED"));
      }

      return Promise.reject(data);
    }

    if (!res.ok) {
      throw new Error(data.message || "Request failed");
    }

    return data;

  } catch (err) {
    console.error("API ERROR:", err.message);
    throw err;
  }
}